Enrich DataRequestEvent MetaAnalytics with varName

samjewell opened the pull request:
DataRequestEvents often come from Panels, but they can also come from
Dashboard Variables that have type "Query".
These events are logged via reportMetaAnalytics. For Panels they are
logged with a `panelId` field present. But for Dashboard Variables
there is no identifier logged, so it's hard to see both:
- Where the data-request was made from (ie. that it came from a dashboard variable at all, and)
- Which dashboard-variable made the data-request

I'd like to add this logging, in order to aid debugging of slow
data-requests coming from dashboard variables. Dashboard-variables can
have more complexity than Panels, because there are different
query-types available which each hit a different Prometheus API endpoint
(in the case of the Prom datasource). As a result, queries can be slow
due to:
- Making a bad choice of query-type for the shape of your data, or
- A regression in one of the Prom API endpoints, for example
   `/label/[label-key]/values`

Having this additional logging would help with debugging these in
general.

These logs are available to all GrafanaCloud customers via the
`grafanacloud-[instance-slug]-usage-insights` Loki datasource, so this
should help our customers too.

See this screenshot of what I'm describing:
![image](https://github.com/grafana/grafana/assets/2903904/af81b672-07be-48b4-9e61-04b7746ab892)

samjewell committed:
Enrich DataRequestEvent MetaAnalytics with varName

DataRequestEvents often come from Panels, but they can also come from
Dashboard Variables that have type "Query".
These events are logged via reportMetaAnalytics. For Panels they are
logged with a `panelId` field present. But for Dashboard Variables
there is no identifier logged, so it's hard to see both:
- Where the data-request was made from (ie. that it came from a dashboard variable at all, and)
- Which dashboard-variable made the data-request

I'd like to add this logging, in order to aid debugging of slow
data-requests coming from dashboard variables. Dashboard-variables can
have more complexity than Panels, because there are different
query-types available which each hit a different Prometheus API endpoint
(in the case of the Prom datasource). As a result, queries can be slow
due to:
- Making a bad choice of query-type for the shape of your data, or
- A regression in one of the Prom API endpoints, for example
   `/label/[label-key]/values`

Having this additional logging would help with debugging these in
general.

These logs are available to all GrafanaCloud customers via the
`grafanacloud-[instance-slug]-usage-insights` Loki datasource, so this
should help our customers too.

See this screenshot of what I'm describing:
![image](https://github.com/grafana/grafana/assets/2903904/af81b672-07be-48b4-9e61-04b7746ab892)

Co-authored-by: Russ <8377044+rdubrock@users.noreply.github.com>

samjewell commented:
> grafana-data has possible breaking changes ([more info](https://github.com/grafana/grafana/actions/runs/5620668232/job/15230211872?check_suite_focus=true#step:8:1))

It seems to me that this is not a breaking change 😕. I'm fairly new to working with Typescript and the Grafana repo, so please let me know if there's something I need to do here.

mckn reviewed (commented):
The code looks good but what do you think about the following:

Mark the `panelId` property as deprecated and replace it with an origin object that looks something like the following:

```ts
origin?: {
   type: 'panel' | 'variable'
   reference: // panelId || variableName
}
```

The benefit with this is that it will scale a bit easier if we would like to add this kind of tracking form e.g. alerting or some other part of the system.

samjewell commented:
@mckn thanks for your review. @gubjanos thought the same in fact, and I can see the value in this.

Is there any chance you will accept this change as it is? And maybe I can follow up in a few weeks or months with the improvement you suggest here? The reason I ask is that:
- I'd like to get this improvement in soon in order to retain momentum: And to follow the change through into a GrafanaCloud insight-logs datasource to check that it's working (and what other types we are missing), and
- I'm very new both to Typescript/React and to the Grafana repo, so expanding the scope is likely to take me a lot of time.

WDYT? Thanks!

mckn commented:
> @mckn thanks for your review. @gubjanos thought the same in fact, and I can see the value in this.
> 
> Is there any chance you will accept this change as it is? And maybe I can follow up in a few weeks or months with the improvement you suggest here? The reason I ask is that:
> 
> * I'd like to get this improvement in soon in order to retain momentum: And to follow the change through into a GrafanaCloud insight-logs datasource to check that it's working (and what other types we are missing), and
> * I'm very new both to Typescript/React and to the Grafana repo, so expanding the scope is likely to take me a lot of time.
> 
> WDYT? Thanks!
> 
> **EDIT:**
> 
> > Mark the panelId property as deprecated
> 
> I'm not sure exactly how I would do this - any thoughts/tips? Altering the log-structure could break some dashboards, right? So the change would go in with a Grafana version-change? I suppose that's an extra reason to make the change you've suggested sooner rather than later: to "rip off the band-aid" as it were.

You can just att the `@deprecated - comment about the replacement` annotation above that property and it should be enough. I'm happy to help you get this change in place during the day.

samjewell committed:
Add a new Origin prop to QueryAnalytics events

This should scale more easily, if we would like to add tracking for eg.
alerting or some other part of the system.

Thanks to @rdubrock and @mckn for pairing and code-review

Co-authored-by: Russ <8377044+rdubrock@users.noreply.github.com>

samjewell commented:
ok @mckn this is ready for a second review now 🙏

dprokop reviewed (changes requested):
Overall it looks good, but the tests are missing coverage for the origin being transported when issuing panel query. I'm talking primarily about this test https://github.com/grafana/grafana/blob/main/public/app/features/query/state/queryAnalytics.test.ts#L126-L127

mckn reviewed (approved):
LGTM!

torkelo commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -521,7 +531,11 @@ export interface DataQueryRequest<TQuery extends DataQuery = DataQuery> {
+   */
   panelId?: number;
+  origin?: DataRequestOrigin;
   dashboardUID?: string;
```
we should depreciate this as well, and move it into DataRequestOrigin

torkelo commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -505,6 +505,16 @@ export interface DataQueryError {
+ *
+ * @public
+ */
+export interface DataRequestOrigin {
+  type: 'panel' | 'variable';
```
Some times queries are not associated with a single panel or variable, we actually have a lot more we want to track

Hm.. maybe something like this 

```ts
export interface DataRequestOrigin {
  queryType: 'panel-data' | 'variable'  | 'annotations' | 'shared';
  panelRef?: string;  // (panel id or could be panel name)
  queryRef?: string;  // Query name / ref for standalone queries, variable name for variable queries
  dashboardUID?: string;  
  pageName?:   // for scene apps plugins
  app?: string;   // the app plugin / or core feature name  (this already exists on the root DataQueryRequest)
}
```

Question what should we set for variable queries?

mckn replied:
I guess we also have alerting as a queryType? And probably more? 

The suggestion with having `type` and some kind of `reference` field was made so we could detect what kind of value is passed in the reference by reading the `type`.

I don't have as good knowledge of which of these values that can exist the same time on the same request, as you do, but it would be nice to end up with a generic structure that could support our current scenarios and hopefully most future scenarios without the need of adding new optional fields.

marefr replied:
Just a random comment from the past in regards to https://github.com/grafana/grafana/pull/11326. Back then already felt "dirty" that datasources have to populate origin information. I feel there's an opportunity to correlate the existing `requestId` with additional context/origin information allowing the  features, e.g. dashboard, explore, alerting, variables to provide the information that make sense to them. This can then be extracted in backend_srv given you have the `requestId`.

torkelo reviewed (changes requested):
I think this is a good change but it needs updates to DataSourcWithBackend (where these are used and sent has headers to the backend so they can be part of the trace & log context). 

https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L77 

There are also backend changes that are needed (and new headers). 

Because requests are now coming from a lot more places (dashboards, explore, app plugins, scene app plugins) etc we need to overhaul these tracking props to better support scene app plugins (and core dashboards). 

```ts
export interface DataRequestOrigin {
  queryType: 'panel-data' | 'variable'  | 'annotations' | 'shared';
  panelRef?: string;  // (panel id or could be panel name)
  queryRef?: string;  // Query name / ref for standalone queries, variable name for variable queries
  dashboardUID?: string;  
  pageName?:   // for scene apps plugins
  app?: string;   // the app plugin / or core feature name  (this already exists on the root DataQueryRequest)
}
``` 

thoughts? 
@dprokop @domasx2 @mckn

mckn commented:
> I think this is a good change but it needs updates to DataSourcWithBackend (where these are used and sent has headers to the backend so they can be part of the trace & log context).
> 
> https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L77
> 
> There are also backend changes that are needed (and new headers).
> 
> Because requests are now coming from a lot more places (dashboards, explore, app plugins, scene app plugins) etc we need to overhaul these tracking props to better support scene app plugins (and core dashboards).
> 
> ```ts
> export interface DataRequestOrigin {
>   queryType: 'panel-data' | 'variable'  | 'annotations' | 'shared';
>   panelRef?: string;  // (panel id or could be panel name)
>   queryRef?: string;  // Query name / ref for standalone queries, variable name for variable queries
>   dashboardUID?: string;  
>   pageName?:   // for scene apps plugins
>   app?: string;   // the app plugin / or core feature name  (this already exists on the root DataQueryRequest)
> }
> ```
> 
> thoughts? @dprokop @domasx2 @mckn

@samjewell  if you feel like the scope of this is starting to grow to much I'm happy to help you move this forward.

samjewell commented:
> @samjewell if you feel like the scope of this is starting to grow to much I'm happy to help you move this forward.

@mckn I'm actually wondering whether you might be happy to take over this work - push it to completion? I no longer really need this change myself, so although I'd like to see this go live, it isn't really a priority for me any more. 
I don't want to lose the work completely, so if you don't have time to take it over I'd happily work together on it to wrap it up. But otherwise it's now just a learning exercise for me. Still valuable, but there's other things I could also be learning.

mckn commented:
> I think this is a good change but it needs updates to DataSourcWithBackend (where these are used and sent has headers to the backend so they can be part of the trace & log context).
> 
> https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L77
> 
> There are also backend changes that are needed (and new headers).
> 
> Because requests are now coming from a lot more places (dashboards, explore, app plugins, scene app plugins) etc we need to overhaul these tracking props to better support scene app plugins (and core dashboards).
> 
> ```ts
> export interface DataRequestOrigin {
>   queryType: 'panel-data' | 'variable'  | 'annotations' | 'shared';
>   panelRef?: string;  // (panel id or could be panel name)
>   queryRef?: string;  // Query name / ref for standalone queries, variable name for variable queries
>   dashboardUID?: string;  
>   pageName?:   // for scene apps plugins
>   app?: string;   // the app plugin / or core feature name  (this already exists on the root DataQueryRequest)
> }
> ```
> 
> thoughts? @dprokop @domasx2 @mckn

Would this be an option to get types that are a bit restricted?

```ts

export type ExploreDataRequestOrigin = {
  type: 'editor' | 'logs';
  reference: string; // panelId or query reference
};

export type DashboardDataRequestOrigin = {
  type: 'panel' | 'annotation' | 'variable' | 'shared';
  reference: string; // variableName or panelId
  dashboardUID: string;
};

export type ScenesDataRequestOrigin = {
  type: 'scenes';
  reference: string; // name of the scenes page
};

export type DataRequestOrigin = ExploreDataRequestOrigin | DashboardDataRequestOrigin | ScenesDataRequestOrigin;


```

mckn commented:
@marefr had another cool idea which could be that we start to send this information as opentelemetry tracing metrics via faro.

mckn commented:
> @marefr had another cool idea which could be that we start to send this information as opentelemetry tracing metrics via faro.

Me and @marefr will look into different ways to get this in place in favour of the suggested solutions in this PR. I will create an issue for it so we can add this to our prioritisation and close this PR. Hope that is OK for everyone 🙏🏻

mckn commented:
Closing in favor of: https://github.com/grafana/grafana/issues/75602
